Sources/NIOCore/AsyncAwaitSupport.swift

```diff
@@ -218,4 +218,120 @@ extension ChannelPipeline {
         try await self.addHandlers(handlers, position: position)
     }
 }
+
+public struct NIOTooManyBytesError: Error {
+     public init() {}
+ }
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension AsyncSequence where Element: RandomAccessCollection, Element.Element == UInt8 {
+    /// Accumulates an ``Swift/AsyncSequence`` of ``Swift/RandomAccessCollection``s into a single `accumulationBuffer`.
+    /// - Parameters:
+    ///   - accumulationBuffer: buffer to write all the elements of `self` into
+    ///   - maxBytes: The maximum number of bytes this method is allowed to write into `accumulationBuffer`
+    /// - Throws: `NIOTooManyBytesError` if the the sequence contains more than `maxBytes`.
+    /// Note that previous elements of `self` might already be write to `accumulationBuffer`.
+    @inlinable
+    public func collect(
+        into accumulationBuffer: inout ByteBuffer,
+        maxBytes: Int
+    ) async throws {
+        var bytesRead = 0
+        for try await fragment in self {
+            bytesRead += fragment.count
+            guard bytesRead <= maxBytes else {
+                throw NIOTooManyBytesError()
+            }
+            accumulationBuffer.writeBytes(fragment)
+        }
+    }
+    
+    /// Accumulates an ``Swift/AsyncSequence`` of ``Swift/RandomAccessCollection``s into a single ``NIO/ByteBuffer``.
+    /// - Parameters:
+    ///   - maxBytes: The maximum number of bytes this method is allowed to accumulate
+    ///   - allocator: Allocator used for allocating the result `ByteBuffer`
+    /// - Throws: `NIOTooManyBytesError` if the the sequence contains more than `maxBytes`.
+    @inlinable
+    public func collect(
+        maxBytes: Int,
+        using allocator: ByteBufferAllocator
```

Comment:
The argument order feels weird to me here (as it is different to all other function calls, where `maxBytes` is the last argument). This is increased through the naming. IIRC plain argument names should come more to the end, whereas argument names that form a sentence should be more at the beginning.

Reply:
Yeah, naming is hard...
I did come up with another idea:
```
func collect(upTo maxBytes: Int, into accumulationBuffer: inout ByteBuffer)
func collect(upTo maxBytes: Int, using allocator: ByteBufferAllocator) -> ByteBuffer
// for the ByteBuffer specific one without allocator
func collect(upTo maxBytes: Int) -> ByteBuffer
```
WDYT?

Reply:
I think I prefer the maxBytes at the end. But that's just my personal opinion. @Lukasa @glbrntt wdyt?

Reply:
I prefer David's suggestion with `upTo maxBytes` first, it reads better.

Reply:
renamed to `upTo maxBytes` and moved it to the front.